### ISSUE TITLE
Feature: add a new assertion that uses assignability checking

### ISSUE THREAD

**Issue by DetachHead** [JoshuaKGoldberg/eslint-plugin-expect-type#18]
types should be compared using the compiler API or something. using a string comparison causes many problems when you aren't sure exactly how the compiler will resolve a type.

some examples:

```
 Expected type to be: 1|2, got: 1 | 2 
```

```ts
type Foo = { foo: number; bar: number }

//fails: Expected type to be: {foo: number; bar: number}, got: Foo
const foo = {} as Foo // $ExpectType {foo: number; bar: number}
```

since typescript is structurally typed, it doesn't make sense for it to behave that way.

**Comment by JoshuaKGoldberg** [comment]
Indeed, this is a tricky one. At first it does seem like a bug... but then knowing that `{foo: number; bar: number}` is the same as `Foo` almost seems like a feature request?

Let's wait to get a bit more feedback. It's not clear to me that we actually would want to disagree with the TypeScript compiler on the exact phrasing of a type _by default_. Maybe a separate command like `$ExpectTypeStructure` would be best? Not sure.

**Comment by danvk** [comment]
Anders handled this in dtslint by supporting multiple alternatives, e.g. `$ExpectType 1|2 || 2|1`.

**Comment by JoshuaKGoldberg** [comment]
Coming back to this:

> checked by the compiler API

There is no compiler API _yet_ for type assignability checking ("is X assignable to Y?"). We're blocked on my favorite TypeScript issue: https://github.com/microsoft/TypeScript/issues/9879

Marking this particular feature request as blocked. We can bikeshed the name, semantics, and any potential default behavior changes of the command once TypeScript provides an assignability API. If it ever does. 😢

**Comment by JoshuaKGoldberg** [comment]
As of https://github.com/microsoft/TypeScript/pull/56448, we can consider this feature unblocked! _(also, https://github.com/typescript-eslint/typescript-eslint/discussions/7936, heh)_

This shouldn't be the _default_ behavior for the plugin. It should be a separate type of assertion. Let's say `$ExpectTypeAssignableTo`?

cc @danvk

**Comment by jakebailey** [comment]
How are you planning on building up the type from just the string? AFAIK there is no way to ask the checker for any type that isn't derived from what was loaded in the program.

(The original issue is one that would be fixed by adopting the union normalization I wrote and is now filed as #108.)

**Comment by JoshuaKGoldberg** [comment]
Hmm that's a good point. Maybe: creating a new source file with almost the same name (or updating the existing one, depending on the API(s)?), setting its text to be similar, and grabbing that type?

**Comment by jakebailey** [comment]
That sounds like a bad time compared to just using an `expectType` function that allows you to write the code in-place like the example above or the libraries linked in the README.

(I would honestly prefer people tend toward writing those and getting feedback directly from `tsc` rather than relying on `ExpectType` comments and was relieved when we ditched `ExpectError` in favor of `// @ts-expect-error` on DT.)

**Comment by JoshuaKGoldberg** [comment]
Makes sense, thanks! I'll close this for now. If anybody has strong opinions they can always post back or file a new issue.

**Comment by JoshuaKGoldberg** [comment]
@all-contributors please add @detachhead for ideas.

> 🤖 Beep boop! This comment was added automatically by [all-contributors-auto-action](https://github.com/marketplace/actions/all-contributors-auto-action).
> Not all contributions can be detected from Git & GitHub alone. Please comment any missing contribution types this bot missed.
> ...and of course, thank you for contributing! 💙
